src/Microsoft.Azure.WebJobs.Host/Loggers/FunctionInstanceLogger.cs

```diff
@@ -24,54 +24,79 @@ public FunctionInstanceLogger(ILoggerFactory loggerFactory)
+            logger.LogInformation(messageTemplate, templateValues);
         }
 
         public Task LogFunctionCompletedAsync(FunctionCompletedMessage message, CancellationToken cancellationToken)
         {
-            if (message.Succeeded)
-            {
-                string traceMessage = string.Format(CultureInfo.InvariantCulture, "Executed '{0}' (Succeeded, Id={1})", message.Function.ShortName, message.FunctionInstanceId);
-                Log(LogLevel.Information, message.Function, message.FunctionInstanceId, traceMessage);
-            }
-            else
-            {
-                string traceMessage = string.Format(CultureInfo.InvariantCulture, "Executed '{0}' (Failed, Id={1})", message.Function.ShortName, message.FunctionInstanceId);
-                Log(LogLevel.Error, message.Function, message.FunctionInstanceId, traceMessage, message.Failure.Exception);
-            }
+            var logger = GetOrCreateLogger(message.Function);
+            Log.FunctionCompleted(
+                logger, 
+                message.Function.ShortName, 
+                message.Succeeded ? "Succeeded" : "Failed", 
+                message.FunctionInstanceId, 
+                message.Failure?.Exception);
 
             return Task.CompletedTask;
         }
 
-        private void Log(LogLevel level, FunctionDescriptor descriptor, Guid functionId, string message, Exception exception = null)
+        public Task DeleteLogFunctionStartedAsync(string startedMessageId, CancellationToken cancellationToken)
         {
-            ILogger logger = _loggerFactory?.CreateLogger(LogCategories.CreateFunctionCategory(descriptor.LogName));
-            logger?.Log(level, 0, message, exception, (s, e) => s);
+            return Task.CompletedTask;
         }
 
-        public Task DeleteLogFunctionStartedAsync(string startedMessageId, CancellationToken cancellationToken)
+        private ILogger GetOrCreateLogger(FunctionDescriptor descriptor)
         {
-            return Task.CompletedTask;
+            return _cachedLoggers.GetOrAdd(descriptor.FullName, _ =>
+            {
+                return _loggerFactory?.CreateLogger(LogCategories.CreateFunctionCategory(descriptor.LogName));
+            });
+        }
+
+        private static class Log
+        {
+            private static readonly Action<ILogger, string, string, Guid, Exception> _functionStarted =
+                LoggerMessage.Define<string, string, Guid>(
+                    LogLevel.Information,
+                    new EventId(1, nameof(FunctionStarted)),
+                    "Executing '{name}' (Reason='{reason}', Id={id})");
+
+            private static readonly Action<ILogger, string, string, Guid, Exception> _functionCompleted =
+                LoggerMessage.Define<string, string, Guid>(
+                    LogLevel.Information,
+                    new EventId(2, nameof(FunctionStarted)),
+                    "Executed '{name}' ({status}, Id={id})");
```

Comment:
Make these `functionName`, `functionResult`, and `invocationId`? Now that we pass the function result as structured data here, we could pull it out into a separate column for the SystemLogger as well for easier querying. @sidkri is already looking into this for adding Duration to this message.

Reply:
Fixed